### PR TITLE
expression, plan: fix the index selection's bug

### LINKED ISSUE #3911: TiDB SQL plan is not correct when using prepare statement with bind parameter not same as field.
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

Execute prepare statement `EXPLAIN SELECT * FROM indextest WHERE date=? ORDER BY success DESC LIMIT 0, 50; ` got a wrong plan when bind parameter with different type of field.

## Table Structure:

```sql
CREATE TABLE `indextest` (
  `name` varchar(255) COLLATE utf8_bin NOT NULL,
  `success` int(11) NOT NULL,
  `date` int(11) NOT NULL,
  KEY `idx_date` (`date`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
```
## Prepare Stmt:

```sql
SELECT * FROM indextest WHERE date=? ORDER BY success DESC LIMIT 0, 50;
```

## Golang sqlx example code:

```go
tidb, err := sqlx.Connect("root@tcp(127.0.0.1:4000)/test")
explainSql := "EXPLAIN SELECT * FROM indextest WHERE date=? ORDER BY success DESC LIMIT 0, 50;"
val := "20170701"
rows, err := tidb.Queryx(explainSql, val)
```

## TiDB Plan:

```
ID: TableScan_9
ParentID: Selection_2
Json: {
    "db": "test",
    "table": "indextest",
    "desc": false,
    "keep order": false,
    "push down info": {
        "limit": 0,
        "access conditions": null,
        "index filter conditions": null,
        "table filter conditions": null
    }
}
----------
ID: Selection_2
ParentID: Sort_4
Json: {
    "condition": [
        "eq(cast(test.indextest.date), 2.0170701e+07)"
    ],
    "scanController": false,
    "child": "TableScan_9"
}
----------
ID: Sort_4
ParentID:
Json: {
    "exprs": [
        {
            "Expr": "indextest.success",
            "Desc": true
        }
    ],
    "limit": 50,
    "child": "Selection_2"
}
```

## MySQL Plan:

```
id: 1
select_type: SIMPLE
table: indextest
type: ref
possible_keys: idx_date
key: idx_date
key_len: 4
ref: const
rows: 1
Extra: Using where; Using filesort
```

2. What did you expect to see?

TiDB Execute Plan should be like this

```
ID: TableScan_8
ParentID: Sort_4
Json: {
    "db": "test",
    "table": "indextest",
    "desc": false,
    "keep order": false,
    "push down info": {
        "limit": 50,
        "sort items": [
            {
                "Expr": "test.indextest.success",
                "Desc": true
            }
        ],
        "access conditions": null,
        "index filter conditions": null,
        "table filter conditions": [
            "eq(test.indextest.date, 20170701)"
        ]
    }
}
----------
ID: Sort_4
ParentID:
Json: {
    "exprs": [
        {
            "Expr": "indextest.success",
            "Desc": true
        }
    ],
    "limit": 50,
    "child": "TableScan_8"
}
```

3. What did you see instead?

4. What version of TiDB are you using (`tidb-server -V`)?

Git Commit Hash: 6b94f4433a0a25655c0bbbcd335a2c615277bcbd
UTC Build Time:  2017-07-24 05:23:17

### PR DESCRIPTION
**Pull request by hanfei1991**
The expression like `a = '1'` or `a < 1.5` can use index from now on.

This should fix #3911

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -294,11 +296,81 @@ func isTemporalColumn(expr Expression) bool {
...
+
+func (c *compareFunctionClass) refineArgs(args []Expression, ctx context.Context) []Expression {
+	arg0IsInt := args[0].GetTypeClass() == types.ClassInt
+	arg1IsInt := args[1].GetTypeClass() == types.ClassInt
+	if (arg0IsInt && arg1IsInt) || (!arg0IsInt && !arg1IsInt) {
+		return args
+	}
+	arg0, arg0IsCon := args[0].(*Constant)
+	arg1, arg1IsCon := args[1].(*Constant)
+	if (!arg0IsInt && !arg0IsCon) || (!arg1IsInt && !arg1IsCon) || (arg0IsCon && arg1IsCon) {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: for my part, the following code is more readable.
``` sql
// int non-constant [cmp] non-int constant
// non-int constant [cmp] int non-constant
if (arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon) ||
   (arg1IsInt && !arg1IsCon && !arg0IsInt && arg0IsCon) {
}
```

File: expression/util.go [hunk 2]
```diff
@@ -298,6 +299,16 @@ var oppositeOp = map[string]string{
 	ast.NE: ast.EQ,
 }
 
+var revertOp = map[opcode.Op]opcode.Op{
```
Comment (XuHuaiyu) [inline expression/util.go]: will symmetricOp be better?

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -294,11 +296,81 @@ func isTemporalColumn(expr Expression) bool {
...
+	if err != nil {
+		return con
+	}
+	return &Constant{
+		Value:   types.NewIntDatum(i64),
+		RetType: types.NewFieldType(mysql.TypeLonglong),
+	}
+}
+
+func refineTwoArgs(con *Constant, op opcode.Op, ctx context.Context) *Constant {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: 1. comment for this func
2. why not make this func be member method of compareFunctionClass.

Comment (hanfei1991) [inline expression/builtin_compare.go]: `op` sometimes is not original.

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -294,11 +296,83 @@ func isTemporalColumn(expr Expression) bool {
...
+}
+
+// refineArgs rewrite the arguments if one of them is int expression and another one is non-int constant.
+// Like a < 1.1 will be changed to a < 2.
+func (c *compareFunctionClass) refineArgs(args []Expression, ctx context.Context) []Expression {
+	arg0IsInt := args[0].GetTypeClass() == types.ClassInt
+	arg1IsInt := args[1].GetTypeClass() == types.ClassInt
+	arg0, arg0IsCon := args[0].(*Constant)
+	arg1, arg1IsCon := args[1].(*Constant)
+	// int non-constant [cmp] non-int constant]
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: remove the useless `]`

File: expression/builtin_compare.go [hunk 5]
```diff
@@ -294,11 +296,83 @@ func isTemporalColumn(expr Expression) bool {
...
+	if err != nil {
+		return con
+	}
+	return &Constant{
+		Value:   types.NewIntDatum(i64),
+		RetType: types.NewFieldType(mysql.TypeLonglong),
+	}
+}
+
+// refineConstantArg change the constant argument to it's ceiling or flooring result by the given op.
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: s/change/changes

**Comment by blacktear23** [comment]
@hanfei1991 will this PR cover Date column compare with String, such as: `created_at > "2017-09-01"` which created_at's type is Date not Datetime.

**Comment by shenli** [comment]
@blacktear23 That case will be fixed in another PR soon.

File: plan/dag_plan_test.go [hunk 6]
```diff
@@ -703,3 +703,240 @@ func (s *testPlanSuite) TestDAGPlanBuilderAgg(c *C) {
...
+			sql:  "select a from t where c_str like 'abc%af'",
+			best: "TableReader(Table(t))->Sel([like(test.t.c_str, abc%af, 92)])->Projection",
+			//			best: "Index(t.c_d_e_str)[[abc <nil>,abd <nil>)]->Selection->Projection",
+		},
+		{
+			sql:  `select a from t where c_str like 'abc\\_' escape ''`,
+			best: "TableReader(Table(t))->Sel([like(test.t.c_str, abc\\_, 92)])->Projection",
+			//			best: "Index(t.c_d_e_str)[[abc_,abc_]]->Projection",
+		},
+		//{
```
Comment (hanfei1991) [inline plan/dag_plan_test.go]: `Like` cann't use index in new planner now.

File: plan/dag_plan_test.go [hunk 7]
```diff
@@ -703,3 +703,240 @@ func (s *testPlanSuite) TestDAGPlanBuilderAgg(c *C) {
...
+			sql:  `select a from t where c > 1.9`,
+			best: "IndexReader(Index(t.c_d_e)[(1,+inf]])->Projection",
+		},
+		{
+			sql:  `select a from t where c = 123456789098765432101234`,
+			best: "TableReader(Table(t))->Sel([eq(cast(test.t.c), 123456789098765432101234)])->Projection",
+		},
+		{
+			sql:  `select a from t where c = 'hanfei'`,
+			best: "TableReader(Table(t))->Sel([eq(cast(test.t.c), cast(hanfei))])->Projection",
```
Comment (winoros) [inline plan/dag_plan_test.go]: It seems that `cast` doesn't be pushed down to tikv, so there should be a selection above the TableReader?

File: expression/builtin_compare_test.go [hunk 8]
```diff
@@ -14,15 +14,78 @@
...
+		{conOne, colInt, ast.GT, "gt(1, a)"},
+		{conOne, colInt, ast.GE, "ge(1, a)"},
+		{conOne, colInt, ast.EQ, "eq(1, a)"},
+		{conOne, colInt, ast.NullEQ, "nulleq(1, a)"},
+		{conOne, colInt, ast.NE, "ne(1, a)"},
+		{conOnePointOne, colInt, ast.LT, "lt(1, a)"},
+		{conOnePointOne, colInt, ast.LE, "le(2, a)"},
+		{conOnePointOne, colInt, ast.GT, "gt(2, a)"},
+		{conOnePointOne, colInt, ast.GE, "ge(1, a)"},
+		{conOnePointOne, colInt, ast.EQ, "eq(1.1, cast(a))"},
```
Comment (XuHuaiyu) [inline expression/builtin_compare_test.go]: why can't this be converted?
cast 1.1 to int will cause truncate error?

**Comment by shenli** [comment]
LGTM

File: expression/builtin_compare_test.go (hunk 8 above)
Comment (hanfei1991) [inline expression/builtin_compare_test.go]: This should be folded as false

File: expression/builtin_compare.go [hunk 9]
```diff
@@ -459,11 +461,83 @@ func isTemporalColumn(expr Expression) bool {
 	return true
 }
 
+func tryToConvertConstantInt(con *Constant, ctx context.Context) *Constant {
```
Comment (zz-jason) [inline expression/builtin_compare.go]: how about let it be a member function of struct `compareFunctionClass` ?

File: expression/builtin_compare.go [hunk 10]
```diff
@@ -459,11 +461,83 @@ func isTemporalColumn(expr Expression) bool {
...
+		return con
+	}
+	return &Constant{
+		Value:   types.NewIntDatum(i64),
+		RetType: types.NewFieldType(mysql.TypeLonglong),
+	}
+}
+
+// refineConstantArg changes the constant argument to it's ceiling or flooring result by the given op.
+func refineConstantArg(con *Constant, op opcode.Op, ctx context.Context) *Constant {
```
Comment (zz-jason) [inline expression/builtin_compare.go]: ditto

File: expression/builtin_compare.go [hunk 11]
```diff
@@ -294,11 +296,83 @@ func isTemporalColumn(expr Expression) bool {
...
+	sc := ctx.GetSessionVars().StmtCtx
+	i64, err := con.Value.ToInt64(sc)
+	if err != nil {
+		log.Warnf("failed to refine this expression, the constant argument is %s", con)
+		return con
+	}
+	datumInt := types.NewIntDatum(i64)
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]:
``` sql
create table t(a bigint unsigned);
select a < cast(-1 as unsigned);
```

`a < cast(-1 as unsigned)` will be refined as `a < -1`?
`a < cast(-1 as unsigned)` will be compared as `unsigned` int,
but `a < -1` will be compared as `signed` int.

Comment (XuHuaiyu) [inline expression/builtin_compare.go]: never mind

File: expression/builtin_compare.go [hunk 12]
```diff
@@ -294,11 +296,83 @@ func isTemporalColumn(expr Expression) bool {
 	return true
 }
 
+func tryToConvertConstantInt(con *Constant, ctx context.Context) *Constant {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: comment for this func, rest LGTM

**Pull request merged**